File: courses/part1-basics/module-1.1-programming-basics/examples/javascript/02_basic_syntax.js
/**
 * 02_basic_syntax.js
 * JavaScript基础语法示例（ES6+）
 */

// 1. 变量声明
console.log("=" .repeat(50));
console.log("1. 变量声明（let和const）");
console.log("=" .repeat(50));

// let - 可变变量
let count = 0;
count = 1;
console.log(`let变量: ${count}`);

// const - 常量（推荐）
const PI = 3.14159;
console.log(`const常量: ${PI}`);

// 注意：const数组/对象可以修改内容
const arr = [1, 2, 3];
arr.push(4);  // 可以修改
console.log(`const数组（可修改内容）: ${arr}`);

// 2. 箭头函数
console.log("\n" + "=" .repeat(50));
console.log("2. 箭头函数");
console.log("=" .repeat(50));

// 传统函数
function traditionalAdd(a, b) {
  return a + b;
}

// 箭头函数（推荐）
const arrowAdd = (a, b) => a + b;
const arrowGreet = (name) => `Hello, ${name}!`;

console.log(`传统函数: ${traditionalAdd(3, 5)}`);
console.log(`箭头函数: ${arrowAdd(3, 5)}`);
console.log(arrowGreet("Alice"));

// 3. 模板字符串
console.log("\n" + "=" .repeat(50));
console.log("3. 模板字符串");
console.log("=" .repeat(50));

const userName = "Bob";
const age = 30;

// 传统拼接
const traditional = "Name: " + userName + ", Age: " + age;

// 模板字符串（推荐）
const template = `Name: ${userName}, Age: ${age}`;

console.log(`传统拼接: ${traditional}`);
console.log(`模板字符串: ${template}`);

// 多行模板字符串
const multiLine = `
第一行
第二行
第三行
`;
console.log("多行字符串:");
console.log(multiLine);

// 4. 解构赋值
console.log("\n" + "=" .repeat(50));
console.log("4. 解构赋值");
console.log("=" .repeat(50));

// 数组解构
const colors = ["red", "green", "blue"];
const [first, second, third] = colors;
console.log(`数组解构: ${first}, ${second}, ${third}`);

// 对象解构
const person = { name: "Alice", age: 25, city: "Beijing" };
const { name, age, city } = person;
console.log(`对象解构: ${name}, ${age}, ${city}`);

// 重命名和默认值
const { name: userName2, job = "Developer" } = person;
console.log(`重命名: ${userName2}, 默认值: ${job}`);

// 5. 扩展运算符
console.log("\n" + "=" .repeat(50));
console.log("5. 扩展运算符");
console.log("=" .repeat(50));

// 数组扩展
const arr1 = [1, 2, 3];
const arr2 = [4, 5, 6];
const combined = [...arr1, ...arr2];
console.log(`数组合并: ${combined}`);

// 对象扩展
const obj1 = { a: 1, b: 2 };
const obj2 = { c: 3, ...obj1 };
console.log("对象合并:", obj2);

// 函数参数
const sum = (...numbers) => numbers.reduce((a, b) => a + b, 0);
console.log(`可变参数求和: ${sum(1, 2, 3, 4, 5)}`);

// 6. 数组方法
console.log("\n" + "=" .repeat(50));
console.log("6. 数组方法");
console.log("=" .repeat(50));

const numbers = [1, 2, 3, 4, 5];

// map - 映射
const doubled = numbers.map(x => x * 2);
console.log(`map（乘2）: ${doubled}`);

// filter - 过滤
const evens = numbers.filter(x => x % 2 === 0);
console.log(`filter（偶数）: ${evens}`);

// reduce - 归约
const total = numbers.reduce((acc, x) => acc + x, 0);
console.log(`reduce（求和）: ${total}`);

// find - 查找
const found = numbers.find(x => x > 3);
console.log(`find（>3）: ${found}`);

// includes - 包含检查
console.log(`includes（包含3）: ${numbers.includes(3)}`);

// 链式调用
const chainResult = numbers
  .filter(x => x % 2 === 0)
  .map(x => x * 10);
console.log(`链式调用: ${chainResult}`);

// 7. 对象操作
console.log("\n" + "=" .repeat(50));
console.log("7. 对象操作");
console.log("=" .repeat(50));

const user = {
  name: "Alice",
  age: 25,
  skills: ["Python", "JavaScript"]
};

// Object.keys/values/entries
console.log("Object.keys:", Object.keys(user));
console.log("Object.values:", Object.values(user));
console.log("Object.entries:", Object.entries(user));

// 遍历对象
console.log("\n遍历对象:");
for (const [key, value] of Object.entries(user)) {
  console.log(`  ${key}: ${value}`);
}

// 8. 字符串方法
console.log("\n" + "=" .repeat(50));
console.log("8. 字符串方法");
console.log("=" .repeat(50));

const text = "Hello, JavaScript World";

console.log(`原文: ${text}`);
console.log(`includes('JavaScript'): ${text.includes('JavaScript')}`);
console.log(`startsWith('Hello'): ${text.startsWith('Hello')}`);
console.log(`endsWith('World'): ${text.endsWith('World')}`);
console.log(`slice(0, 5): ${text.slice(0, 5)}`);
console.log(`split(','): ${text.split(',')}`);
console.log(`replace('JavaScript', 'AI'): ${text.replace('JavaScript', 'AI')}`);

// 9. JSON操作
console.log("\n" + "=" .repeat(50));
console.log("9. JSON操作");
console.log("=" .repeat(50));

// 对象转JSON字符串
const jsonObj = { name: "Alice", age: 25 };
const jsonString = JSON.stringify(jsonObj);
console.log(`对象转JSON: ${jsonString}`);

// 格式化JSON
const formatted = JSON.stringify(jsonObj, null, 2);
console.log("格式化JSON:");
console.log(formatted);

// JSON字符串转对象
const parsed = JSON.parse(jsonString);
console.log("JSON转对象:", parsed);

console.log("\n示例完成!");

File: courses/part1-basics/module-1.1-programming-basics/examples/python/02_basic_syntax.py
"""
02_basic_syntax.py
Python基础语法示例
"""

# 1. 变量与数据类型
print("=" * 50)
print("1. 变量与数据类型")
print("=" * 50)

# 数字类型
integer_num = 42
float_num = 3.14159
complex_num = 2 + 3j

print(f"整数: {integer_num}, 类型: {type(integer_num)}")
print(f"浮点数: {float_num}, 类型: {type(float_num)}")
print(f"复数: {complex_num}, 类型: {type(complex_num)}")

# 字符串
single_quote = 'Hello'
double_quote = "World"
f_string = f"{single_quote} {double_quote}"

print(f"字符串拼接: {f_string}")
print(f"字符串长度: {len(f_string)}")

# 布尔值
is_true = True
is_false = False

print(f"布尔值: {is_true}, {is_false}")

# 2. 数据结构
print("\n" + "=" * 50)
print("2. 数据结构")
print("=" * 50)

# 列表
fruits = ["apple", "banana", "cherry", "date"]
print(f"原始列表: {fruits}")

fruits.append("elderberry")
print(f"添加后: {fruits}")

print(f"切片 [1:3]: {fruits[1:3]}")
print(f"最后两个: {fruits[-2:]}")

# 字典
person = {
    "name": "Alice",
    "age": 25,
    "city": "Beijing"
}
print(f"字典: {person}")
print(f"姓名: {person['name']}")
print(f"年龄: {person.get('age')}")

# 集合
colors = {"red", "green", "blue", "red"}  # 自动去重
print(f"集合（自动去重）: {colors}")

# 元组
coordinates = (10, 20)
print(f"元组: {coordinates}")
x, y = coordinates  # 解包
print(f"解包: x={x}, y={y}")

# 3. 运算符
print("\n" + "=" * 50)
print("3. 运算符")
print("=" * 50)

a, b = 10, 3

print(f"a + b = {a + b}")  # 加法
print(f"a - b = {a - b}")  # 减法
print(f"a * b = {a * b}")  # 乘法
print(f"a / b = {a / b}")  # 除法
print(f"a // b = {a // b}")  # 整除
print(f"a % b = {a % b}")  # 取模
print(f"a ** b = {a ** b}")  # 幂运算

# 比较运算
print(f"a > b: {a > b}")
print(f"a == b: {a == b}")

# 4. 控制流
print("\n" + "=" * 50)
print("4. 控制流")
print("=" * 50)

# if语句
score = 85
if score >= 90:
    grade = "A"
elif score >= 80:
    grade = "B"
elif score >= 70:
    grade = "C"
else:
    grade = "D"

print(f"分数 {score}, 等级: {grade}")

# for循环
print("遍历列表:")
for i, fruit in enumerate(fruits[:3]):
    print(f"  {i}: {fruit}")

# while循环
print("\n计数循环:")
count = 0
while count < 3:
    print(f"  计数: {count}")
    count += 1

# 5. 字符串操作
print("\n" + "=" * 50)
print("5. 字符串操作")
print("=" * 50)

text = "Hello, Python World"

print(f"原文: {text}")
print(f"小写: {text.lower()}")
print(f"大写: {text.upper()}")
print(f"标题: {text.title()}")
print(f"替换: {text.replace('Python', 'AI')}")
print(f"分割: {text.split(',')}")

# 6. 列表推导式
print("\n" + "=" * 50)
print("6. 列表推导式")
print("=" * 50)

numbers = [1, 2, 3, 4, 5]

# 基本推导式
squares = [x ** 2 for x in numbers]
print(f"平方: {squares}")

# 条件推导式
evens = [x for x in numbers if x % 2 == 0]
print(f"偶数: {evens}")

# 字典推导式
squared_dict = {x: x ** 2 for x in numbers}
print(f"平方字典: {squared_dict}")

print("\n示例完成!")